File: src/tokenizer.cpp
#include "language/tokenizer.h"

#include <cctype>

Token Tokenizer::scan_token()
{
    skip_whitespace();
    
    start = current;
    
    if (is_at_end()) return make_token(TokenType::Eof);
    
    auto c = advance();
    
    if (isdigit(c)) return number();
    if (isalpha(c)) return identifier();
    
    switch (c) {
        case '(': return make_token(TokenType::OpenParen);
        case ')': return make_token(TokenType::CloseParen);
        case '[': return make_token(TokenType::OpenSquare);
        case ']': return make_token(TokenType::CloseSquare);
        case '{': return make_token(TokenType::OpenCurly);
        case '}': return make_token(TokenType::CloseCurly);
        case ';': return make_token(TokenType::Semicolon);
        case ',': return make_token(TokenType::Comma);
        case '.': return make_token(TokenType::Dot);
        case '-': return make_token(TokenType::Minus);
        case '+': return make_token(TokenType::Plus);
        case '/': return make_token(TokenType::Slash);
        case '*': return make_token(TokenType::Star);
        case '^': return make_token(TokenType::BwXor);
        case '&': return make_token(match('&') ? TokenType::And : TokenType::BwAnd);
        case '|': return make_token(match('|') ? TokenType::Or : TokenType::BwOr);
        case '!':
            return make_token(match('=') ? TokenType::ExclEqual : TokenType::Excl);
        case '=':
            return make_token(match('=') ? TokenType::EqualEqual : TokenType::Equal);
        case '<':
            return make_token(match('=') ? TokenType::LessEqual : match('<') ?
                              TokenType::LessLess : TokenType::Less);
        case '>':
            return make_token(match('=') ? TokenType::GreaterEqual : match('>') ?
                              TokenType::GreaterGreater : TokenType::Greater);
            
        case '"': return string_();
        case '\'': return string_('\'');
        default: return error_token("Unexpected character.");
    }
}

bool Tokenizer::is_at_end() const
{
    return current == source.length();
}

char Tokenizer::advance()
{
    current++;
    return source[current - 1];
}

char Tokenizer::peek()
{
    if (is_at_end()) return '\0';
    return source[current];
}

char Tokenizer::peek_next()
{
    if ((current + 1) >= source.length()) return '\0';
    return source[current + 1];
}

bool Tokenizer::match(char expected)
{
    if (is_at_end()) return false;
    if (source[current] != expected) return false;
    
    current++;
    return true;
}

Token Tokenizer::make_token(TokenType type)
{
    auto text = std::string_view(&source[start], current - start);
    return Token(type, text, line);
}

Token Tokenizer::error_token(const std::string& message) const
{
    return Token(TokenType::Error, message, line);
}

void Tokenizer::skip_whitespace()
{
    while (true) {
        auto c = peek();
        switch (c) {
            case ' ':
            case '\r':
            case '\t':
                advance();
                break;
                
            case '\n':
                line++;
                advance();
                break;
                
            case '/':
                if (peek_next() != '/') return;
                while (peek() != '\n' && !is_at_end()) advance();
                break;
                
            default:
                return;
        }
    }
}

TokenType Tokenizer::check_keyword(size_t pos, size_t len, const std::string_view& rest, TokenType type) const
{
    if (current - start == pos + len && source.compare(start + pos, len, rest) == 0) {
        return type;
    }
    
    return TokenType::Identifier;
}

TokenType Tokenizer::identifier_type()
{
    switch (source[start]) {
        case 'c': return check_keyword(1, 4, "lass", TokenType::Class);
        case 'e': return check_keyword(1, 3, "lse", TokenType::Else);
        case 'f':
            if (current - start > 1) {
                switch (source[start + 1]) {
                    case 'a': return check_keyword(2, 3, "lse", TokenType::False);
                    case 'o': return check_keyword(2, 1, "r", TokenType::For);
                    case 'u': return check_keyword(2, 2, "nc", TokenType::Func);
                    default: return TokenType::Identifier;
                }
            }
            break;
        case 'i': return check_keyword(1, 1, "f", TokenType::If);
        case 'n': return check_keyword(1, 3, "ull", TokenType::Null);
        case 'p': return check_keyword(1, 4, "rint", TokenType::Print);
        case 'r': return check_keyword(1, 5, "eturn", TokenType::Return);
        case 's': return check_keyword(1, 4, "uper", TokenType::Super);
        case 't':
            if (current - start > 1) {
                switch (source[start + 1]) {
                    case 'h': return check_keyword(2, 2, "is", TokenType::This);
                    case 'r': return check_keyword(2, 2, "ue", TokenType::True);
                    default: return TokenType::Identifier;
                }
            }
            break;
        case 'v': return check_keyword(1, 2, "ar", TokenType::Var);
        case 'w': return check_keyword(1, 4, "hile", TokenType::While);
        default: return TokenType::Identifier;
    }

    return TokenType::Identifier;
}

Token Tokenizer::identifier()
{
    while (isalnum(peek())) advance();
    
    return make_token(identifier_type());
}

Token Tokenizer::number()
{
    while (isdigit(peek())) advance();
    
    if (peek() == '.' && isdigit(peek_next())) {
        advance();
        while (isdigit(peek())) advance();
    }
    
    return make_token(TokenType::Number);
}

Token Tokenizer::string_(const char open_char)
{
    while (peek() != open_char && !is_at_end()) {
        if (peek() == '\n') line++;
        advance();
    }
    
    if (is_at_end()) return error_token("Unterminated string.");
    
    // The closing ".
    advance();
    return make_token(TokenType::String);
}

File: src/language/tokenizer.rs
use super::token::{Token, TokenType};


/// Used to tokenize (transform into single tokens) source code
/// in the Elysabettian programming language
pub struct Tokenizer {
    source: String,
    start: usize,
    current: usize,
    line: i32
}

impl Tokenizer {
    /// Create a new Tokenizer for this block of code
    /// 
    /// Arguments:
    /// * `source`: reference to source code string
    pub fn new(source: &String) -> Tokenizer {
        Tokenizer {
            source: source.clone(),
            start: 0_usize,
            current: 0_usize,
            line: 1_i32
        }
    }

    pub fn get_token(&mut self) -> Token {
        // Skip all whitespaces
        self.skip_whitespaces();

        self.start = self.current;

        // Check for EOF
        if self.is_at_end() {
            return self.make_token(TokenType::Eof)
        }

        // Get the next character and identify it
        let c = self.advance();
        if c.is_numeric() { return self.number() }
        if c.is_alphanumeric() { return self.identifier() }

        match c {
            '(' => self.make_token(TokenType::OpenParen),
            ')' => self.make_token(TokenType::CloseParen),
            '[' => self.make_token(TokenType::OpenSquare),
            ']' => self.make_token(TokenType::CloseSquare),
            '{' => self.make_token(TokenType::OpenCurly),
            '}' => self.make_token(TokenType::CloseCurly),
            ';' => self.make_token(TokenType::Semicolon),
            ',' => self.make_token(TokenType::Comma),
            '.' => self.make_token(TokenType::Dot),
            '-' => self.make_token(TokenType::Minus),
            '+' => self.make_token(TokenType::Plus),
            '/' => self.make_token(TokenType::Slash),
            '*' => self.make_token(TokenType::Star),
            '^' => self.make_token(TokenType::BwXor),
            '&' => self.make_token(if self.match_token('&') { TokenType::And} else { TokenType::BwAnd}),
            '|' => self.make_token(if self.match_token('|') { TokenType::Or } else { TokenType::BwOr }),
            '!' => self.make_token(if self.match_token('=') { TokenType::ExclEqual } else { TokenType::Excl }),
            '<' => self.make_token(
                if self.match_token('=') 
                { 
                    TokenType::LessEqual
                } else if self.match_token('<'){ 
                    TokenType::LessLess
                })),
            '>' => self.make_token(if self.match_token('=') { TokenType::GreaterEqual } else {})
        }
    }

    /// Checks if the tokenizer has reached the end of the source code
    fn is_at_end(&self) -> bool {
        self.current == self.source.len()
    }

    /// Advances the tokenizer by one character
    fn advance(&mut self) -> char {
        self.current += 1;
        self.source.chars().nth(self.current - 1).unwrap()
    }

    /// Returns the character in the current position
    fn peek(&self) -> char {
        if self.is_at_end() {
            return '\0'
        }

        self.source.chars().nth(self.current).unwrap()
    }

    /// Returns the character in the next position
    fn peek_next(&self) -> char {
        if self.current + 1 >= self.source.len() {
            return '\0'
        }

        self.source.chars().nth(self.current + 1).unwrap()
    }

    /// Matches the token with the current character and advances by one position
    fn match_token(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false
        }

        if self.source.chars().nth(self.current).unwrap() != expected {
            return false
        }

        self.current += 1;
        true
    }

    /// Creates a new token with chars at current position and given type
    fn make_token(&self, token_type: TokenType) -> Token {
        let text: String = self.source.chars()
            .skip(self.start)
            .take(self.current - self.start)
            .collect();

        Token::new(token_type, &text, self.line)
    }

    /// Creates a new Token with an error message and the corresponding line
    fn error_token(&self, message: &str) -> Token {
        Token::new(TokenType::Error, &message.to_owned(), self.line)
    }

    /// Matches all whitespaces and skips them if necessary
    fn skip_whitespaces(&mut self) {
        loop {
            // Take the current character
            let c = self.peek();

            // Match it against all whitespaces and skip them if necessary
            match c {
                ' ' | '\r' | '\t' => _ = self.advance(),
                '\n' => { self.line += 1; _= self.advance() },
                '/' => {
                    if self.peek_next() != '/' {
                        return;
                    }

                    while self.peek() != '\n' && !self.is_at_end() {
                        _ = self.advance();
                    }
                },
                _ => return
            }
        }
    }

    /// Checks a keyword and, if it is a language keyword, it sets its corresponding token type
    fn check_keyword(&self, pos: usize, len: usize, rest: &str, token_type: TokenType) -> TokenType {
        let partial_text: String = self.source.chars()
            .skip(self.start + pos)
            .take(len)
            .collect();

        if self.current - self.start == pos + len && rest == partial_text {
            return token_type
        }

        TokenType::Identifier
    }

    /// Finds the type of the given identifier string
    fn identifier_type(&mut self) -> TokenType {
        // Get current character
        let char = self.source.chars().nth(self.start).unwrap();

        // Match it with a list of keyword beginnings
        match char {
            'c' => self.check_keyword(1, 4, "lass", TokenType::Class),
            'e' => self.check_keyword(1, 3, "lse", TokenType::Else),
            'f' => {
                if (self.current - self.start > 1) {
                    let m_char = self.source.chars().nth(self.start + 1).unwrap();
                    match m_char {
                        'a' => self.check_keyword(2, 3, "lse", TokenType::False),
                        'o' => self.check_keyword(2, 1, "r", TokenType::For),
                        'u' => self.check_keyword(2, 2, "nc", TokenType::Func),
                        _ => TokenType::Identifier
                    }
                } else {
                    TokenType::Identifier
                }
            },
            'i' => self.check_keyword(1, 1, "f", TokenType::If),
            'n' => self.check_keyword(1, 3, "ull", TokenType::Null),
            'p' => self.check_keyword(1, 4, "rint", TokenType::Print),
            'r' => self.check_keyword(1, 5, "eturn", TokenType::Return),
            's' => self.check_keyword(1, 4, "uper", TokenType::Super),
            't' => {
                if self.current - self.start > 1 {
                    let m_char = self.source.chars().nth(self.start + 1).unwrap();
                    match m_char {
                        'h' => self.check_keyword(2, 2, "is", TokenType::This),
                        'r' => self.check_keyword(2, 2, "ue", TokenType::True),
                        _ => TokenType::Identifier
                    }
                } else {
                    TokenType::Identifier
                }
            },
            'v' => self.check_keyword(1, 2, "ar", TokenType::Var),
            'w' => self.check_keyword(1, 4, "hile", TokenType::While),
            _ => TokenType::Identifier
        }
    }

    /// Creates a new identifier token
    fn identifier(&mut self) -> Token {
        while self.peek().is_alphanumeric() {
            _ = self.advance();
        }

        let token_type = self.identifier_type();
        self.make_token(token_type)
    }

    /// Creates a new number token
    fn number(&mut self) -> Token {
        while self.peek().is_numeric() {
            _ = self.advance()
        }

        if self.peek() == '.' && self.peek_next().is_numeric() {
            _ = self.advance();
            while self.peek().is_numeric() {
                _ = self.advance();
            }
        }

        self.make_token(TokenType::Number)
    }

    /// Creates a new string token
    fn string(&mut self, open_char: char) -> Token {
        while self.peek() != open_char && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            
            _ = self.advance();
        }

        if self.is_at_end() {
            return self.error_token("Unterminated string literal!")
        }

        // Closing string delimiter
        _ = self.advance();
        self.make_token(TokenType::String)
    }
}
